cpp/cuproj/benchmarks/wgs_to_utm_bench.cu

```diff
@@ -0,0 +1,130 @@
+ */
+
+#include <cuproj/projection_factories.hpp>
+
+#include <cuproj_test/convert_coordinates.hpp>
+#include <cuproj_test/coordinate_generator.cuh>
+
+#include <benchmarks/fixture/benchmark_fixture.hpp>
+#include <benchmarks/synchronization/synchronization.hpp>
+
+#include <cuspatial/geometry/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_vector.hpp>
+
+#include <thrust/host_vector.h>
+
+#include <type_traits>
+
+template <typename T>
+using coordinate = typename cuspatial::vec_2d<T>;
+
+static char const* epsg_src = "EPSG:4326";
+static char const* epsg_dst = "EPSG:32756";
+
+template <typename T>
+auto make_input(std::size_t grid_side)
+{
+  // Sydney Harbour
+  coordinate<T> min_corner{-33.9, 151.2};
+  coordinate<T> max_corner{-33.7, 151.3};
+
+  auto input = cuproj_test::make_grid_array<coordinate<T>, rmm::device_vector<coordinate<T>>>(
+    min_corner, max_corner, grid_side, grid_side);
+
+  return input;
+}
+
+template <typename T>
+static void cuproj_wgs_to_utm_benchmark(benchmark::State& state)
+{
+  auto const grid_side{static_cast<std::size_t>(state.range(0))};
+
+  auto input = make_input<T>(grid_side);
+
+  rmm::device_vector<coordinate<T>> output(input.size());
+
+  auto proj = cuproj::make_projection<coordinate<T>>(epsg_src, epsg_dst);
+
+  for (auto _ : state) {
+    cuda_event_timer raii(state, true);
+    proj.transform(input.begin(),
+                   input.end(),
+                   output.begin(),
+                   cuproj::direction::FORWARD,
+                   rmm::cuda_stream_default);
+  }
+
+  state.SetItemsProcessed(grid_side * grid_side * state.iterations());
+}
+
+void proj_wgs_to_utm_benchmark(benchmark::State& state)
+{
+  using T = double;
+  auto const grid_side{static_cast<std::size_t>(state.range(0))};
+
+  auto d_input = make_input<T>(grid_side);
+  auto input   = thrust::host_vector<coordinate<T>>(d_input);
+
+  std::vector<PJ_COORD> pj_input(input.size());
+  cuproj_test::convert_coordinates(input, pj_input);
+
+  PJ_CONTEXT* C = proj_context_create();
+  PJ* P         = proj_create_crs_to_crs(C, epsg_src, epsg_dst, nullptr);
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    cuproj_test::convert_coordinates(input, pj_input);
+    state.ResumeTiming();
+    proj_trans_array(P, PJ_FWD, pj_input.size(), pj_input.data());
+  }
+
+  state.SetItemsProcessed(grid_side * grid_side * state.iterations());
+}
+
+class proj_utm_benchmark : public ::benchmark::Fixture {};
+
+BENCHMARK_DEFINE_F(proj_utm_benchmark, forward_double)(::benchmark::State& state)
+{
+  proj_wgs_to_utm_benchmark(state);
+}
+BENCHMARK_REGISTER_F(proj_utm_benchmark, forward_double)
+  ->Range(8, 16384)
+  ->Unit(benchmark::kMillisecond);
+
+class cuproj_utm_benchmark : public cuspatial::benchmark {
+  void SetUp(const ::benchmark::State& state) override
+  {
+    mr = std::make_shared<rmm::mr::cuda_memory_resource>();
+    rmm::mr::set_current_device_resource(mr.get());  // set default resource to cuda
```

Comment:
So the benchmarks are run on cuda mr? What about the pool mr in the parent class `SetUp`?

Reply:
For some reason I'm getting an OOM with the fixture pool MR. Need to debug.

Reply:
There was an OOM error, so I switched the pool off. But just realized that this was because the fixture was not deleting the pool in TearDown. Fixing.

Reply:
Pool MR definitely makes a difference, especially for smaller data sizes and fp32.